qiskit/circuit/quantumcircuit.py

```diff
@@ -1486,6 +1486,85 @@ def _update_parameter_table(self, instruction: CircuitInstruction):
                     # clear cache if new parameter is added
                     self._parameters = None
 
+    @typing.overload
+    def get_parameter(self, name: str, default: T) -> Union[Parameter, T]:
+        ...
+
+    # The builtin `types` module has `EllipsisType`, but only from 3.10+!
+    @typing.overload
+    def get_parameter(self, name: str, default: type(...) = ...) -> Parameter:
```

Comment:
I'm wondering whether it would be cleaner to have a designated object to indicate no argument has been given, personally I stumbled a bit over the Ellipsis since I've only known it in context of slicing or repeating something. Alternatively we could also provide two `get_parameter`, once with the `default` and once without, it seems like that's how Python's `getattr` is implemented.

Reply:
This function has a counterpart `get_var` that was approved in Qiskit/RFCs#50, so this function should probably feel similar to that.

The built-in `getattr` is actually done with positional-only arguments, but I had loosely intended `default=None` to be a valid spelling, and `None` is a very common choice for the default return value, so it needs to be a different object.  The reason for a literal `Ellipsis` is copied from `get_var`; the value appears in the rendered documentation, and I just wanted a simple shorthand so that something that _reads_ as "no value" appears there.

Some libraries make a specific object
```python
class _NotGiven:
    """Type of a sentinel object for an unsupplied argument."""

    __slots__ = ()

    def __repr__(self):
        return "<not given>"

    __str__ = __repr__

_NOT_GIVEN = _NotGiven()
```
or the like for that purpose, so I can switch to that if you prefer.  The downside is that it's like 10 lines of code, and you have to reference the `_NOT_GIVEN` name, and so on, and if we're using the same pattern in more than one file, we either have to rewrite it or re-import it, and so on.  Using `...` since it's just built-in was a shorthand.

Reply:
> The built-in `getattr` is actually done with positional-only arguments

To explain: it's implemented in C, but in Python-space code, the logic looks something like
```python
def getattr(*args):
    nargs = len(args)
    if not 2 <= nargs <= 3:
        raise TypeError(...)
    obj = args[0]
    name = args[1]
    try:
        return <builtin getattr>(obj, name)
    except AttributeError:
        if nargs == 3:
            return args[2]
        raise
```
but that's obviously ugly and slow in Python space.  Easier (and plays nicer with code introspection) to use an explicit call signature, rather than a `*args`.

Reply:
Ah yes, we'd have to be consistent with `get_var`. I think it'd be a bit easier to read with a separate class but with the comment in the `get_parameter` implementation that states why an `Ellipsis` is used I'm good 🙂